src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,196 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
-    @Override
-    public void add(T value) {
+    private static final int DEFAULT_CAPACITY;
+    private static final int SOFT_MAX_ARRAY_LENGTH;
+    private static final Object[] DEFAULT_CAPACITY_EMPTY_ELEMENT_DATA;
+    private static final int NOT_EXISTING_INDEX;
 
+    static {
+        NOT_EXISTING_INDEX = -1;
+        DEFAULT_CAPACITY = 10;
+        SOFT_MAX_ARRAY_LENGTH = Integer.MAX_VALUE - 8;
+        DEFAULT_CAPACITY_EMPTY_ELEMENT_DATA = new Object[]{};
     }
 
-    @Override
-    public void add(T value, int index) {
+    private Object[] elementData;
+    private int size;
 
+    public ArrayList() {
+        elementData = DEFAULT_CAPACITY_EMPTY_ELEMENT_DATA;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        Object[] objectsList = toArray(list);
+        int numNew = objectsList.length;
+        if (numNew == 0) {
+            return;
+        }
+        Object[] elementData;
+        final int size;
+        if (numNew > (elementData = this.elementData).length - (size = this.size)) {
+            elementData = grow(size + numNew);
+        }
+        System.arraycopy(objectsList, 0, elementData, size, numNew);
+        this.size = size + numNew;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, size);
+        return elementData(index);
     }
 
     @Override
     public void set(T value, int index) {
+        checkIndex(index, size);
+        elementData[index] = value;
+    }
 
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return size == 0;
+    }
+
+    public Object[] toArray(List<T> list) {
+        Object[] objects = new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            objects[i] = list.get(i);
+        }
+        return objects;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, size);
+        final Object[] elements = elementData;
+        @SuppressWarnings("unchecked")
+        T oldValue = (T) elements[index];
+        fastRemove(elements, index);
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int searchingIndex = foundElementIndex(element);
+        if (searchingIndex > NOT_EXISTING_INDEX) {
+            fastRemove(elementData, searchingIndex);
+            return element;
+        }
+        throw new NoSuchElementException("Not found such element.");
+    }
+
+    private int foundElementIndex(T element) {
```

Comment:
```suggestion
    private int findIndex(T element) {
```
Also let's move your private methods after public ones